### ISSUE TITLE
Make Jetty JPMS compliant

### ISSUE THREAD

**Issue by PashaTurok** [eclipse/jetty.project#2189]
I suggest to make Jetty fully JPMS compliant as JPMS gives a lot of advantages. Besides I suggest to do it as soon as possible without waiting Java 11, 12 etc.

**Comment by sbordet** [comment]
> JPMS gives a lot of advantages

Which ones are you interested in ?

**Comment by PashaTurok** [comment]
@sbordet Sorry, don't understand your question, because JPMS Jetty will be used not only by me. So it is not important what I (one of thousand users) need.

**Comment by sbordet** [comment]
Well, you are the first one to ask this here, so we are interested in your use case.
What is the advantage that you would have ?
What is it that you would be able to do if Jetty was JPMS compliant that you cannot already do ?
Answers to these questions from the community will help us drive the effort.

**Comment by PashaTurok** [comment]
@sbordet Sorry, I really don't understand you. I can't understand situation when someone opens an issue and suggests new feature and he is asked about his private usage of this new feature. I think that such talking has no sense. It is like - Where is toilet? Why do you need it?

**Comment by sbordet** [comment]
Duplicate of #2127.

**Comment by joakime** [comment]
@sbordet Issue #2127 is not about making Jetty JPMS compliant.

Issue #2127 is about making the automatic names applied to non-JPMS jars more sane in the case of JPMS users wanting to use Jetty.

Issue #2127 is about the name of the JAR, and only the name of the JAR.  

To be JPMS compliant we have package structure changes, artifact changes, metadata additions, manifest changes, ServiceLoader changes, bytecode scanning changes, JEP238 support issues, etc ...

**Comment by joakime** [comment]
The OSGi layers have to be evaluated too.
Either we pick a solution that allows both OSGi and/or JPMS usage, or go 100% JPMS and let the OSGi users use the JPMS compatibility layers

**Comment by PashaTurok** [comment]
@joakime I think that Jetty must support OSGi till certain version, for example till Jetty 10, but later must support only JPMS. Supporting two modular systems - JPMS and OSGi at the same time seems to be bad idea because I am sure that OSGi won't live after Java 8. The problem is that OSGi is very dynamic - it is dynamic per bundle, JPMS is dynamic per layer - in JPMS if layer is created (with its JPMS modules) it can be modified any more - I asked questions about it - see https://mail.osgi.org/pipermail/osgi-dev/2017-September/006548.html and https://stackoverflow.com/questions/46112809/is-it-possible-to-load-and-unload-jdk-and-custom-modules-dynamically-in-java-9.

**Comment by jmcc0nn3ll** [comment]
It is not that we don't want to add JPMS support, it has been a project topic for quite some time.  @joakime is spot on regarding OSGi though and part of our waiting on it has been to see how that community responds to these changes.  @PashaTurok could be correct in that OSGi is irrelevant moving forward but frankly, OSGi has been around a heck of a lot longer then JPMS and while my personal opinions on OSGi are not fit to print..one does not simply drop support and transition on something like this based on one user's issue.  Fundamentally this is why @sbordet asked you about your use case, not out of trying to mine out what you are up to, but because you are legitimately the first person that has approached us and asked for this support...so we would of course like to understand your general use case and expectations of what JPMS support in Jetty would look like.  We knew it would happen, you simply won the first prize of being asked some questions. :)

**Comment by PashaTurok** [comment]
@jmcc0nn3ll Ok. I will explain what is in my mind. JPMS like OSGi is a modular system. If we want any product (in our case Jetty) work in certain modular system we must modify our product according to specs/requirements of this modular system - build modules, hide private packages, implement module services etc. It doesn't matter what I (one of 1000 users) want/need. If a decision to make Jetty JPMS compliant is taken there is only one way - using specs make modular product. I really tried to express my idea.

**Comment by joakime** [comment]
Some Links about the OSGi <-> JPMS compatibilty concern.

* [Letter to our OSGi User Community RE: the Java Platform Module System (JPMS)](http://blog.osgi.org/2017/09/letter-to-our-osgi-user-community-re.html) - _Sept 2017_
* [Eclipse Equinox with Java Modules All the Way Down](https://www.eclipse.org/community/eclipse_newsletter/2016/october/article3.php) - _Oct 2017_

**Comment by jmcc0nn3ll** [comment]
@PashaTurok we get that, i am more interested in answers related to things like are you shooting for a deployment where you use it embedded and manually registering servlets, as a static content server, deploying with the distribution, embedded using webapp deployment via isolated WebappClassloader, planned logging methodologies, http2, websocket, etc.  We have had a working understanding of how it would all play out for quite some time, but that is from the project perspective.  User stories are useful to validate decisions.

**Comment by joakime** [comment]
User Stories are very valuable to us.
It helps frame decisions we have to make during the effort to make Jetty JPMS compliant.

**Comment by PashaTurok** [comment]
@jmcc0nn3ll I thought about these questions and this is what I think.
1) There must be a framework that creates JPMS layers and deploys there (using xml descriptor) certain JPMS modules. This is a lowest level framework.
2) When Jetty becomes a set of JPMS modules, framework will run Jetty in separate layer.
3) For every web site (every site is a set of JPMS modules) framework creates new layer, that is the child of Jetty layer. This is the point where I stopped.
4) About logging - I already modified slf4j to support logging per layer. Maybe it will be necessary to modify and other logging frameworks. As every layer is separate application (Jetty, web site) it is possible to work with certain messages.
Something like this. Many ideas I took from Karaf, pax-logging, pax-web.

**Comment by jmcc0nn3ll** [comment]
Thanks for the feedback @PashaTurok, I have opened up a general issue for this #2191 and reached out to some others to chime in there.  I'll close this in favor of that thread but I have called back to this thread so we don't lose the context.
